[Dashborad] Fix the bugs in the loop in job_manager that checks for Pending status.

daiping8 opened the pull request:
## Overview

To fix Issue #55133

## Problem Description

In the `_monitor_job_internal` method, the code includes a critical ping check:
```python
await job_supervisor.ping.remote()
```

This ping serves several intended purposes:
1. If the job supervisor has been successfully scheduled and created, and the job is in `RUNNING` or `PENDING` state, the ping succeeds and the monitoring loop continues normally.
2. If the cluster has sufficient resources to schedule the supervisor but creation fails (e.g., due to RuntimeEnv issues), the ping will detect the exception and update the job status to `FAILED`.

However, there's a critical edge case:
- **(Exceptional scenario)** When the cluster lacks sufficient resources to schedule the job supervisor, leaving the job in `PENDING` state, the ping operation can block indefinitely. This prevents the job from timing out properly as intended.

The blocking behavior occurs because the ping request remains unresolved while the supervisor actor is stuck in a pending scheduling state, effectively preventing the timeout logic from triggering within `DEFAULT_JOB_START_TIMEOUT_SECONDS`. This creates a scenario where jobs can get permanently stuck in `PENDING` rather than transitioning to `FAILED` after the timeout threshold is reached.


## Solutions

https://github.com/ray-project/ray/blob/94e8d02a9e7f19227ea6d2a4a1e2cb09db55e760/python/ray/dashboard/modules/job/job_manager.py#L189-L234

We should add the two lines of code L233-L234. we should `continue` so it goes back to the beginning of the loop and try again, instead of let it run the remaining code which has `job_supervisor.ping.remote()` which will hang forever.       

## Results
After applying this change, jobs that remain in the `PENDING` state will transition to the `FAILED` state correctly after the timeout is reached.

<img width="2512" height="445" alt="image" src="https://github.com/user-attachments/assets/54af68bc-9152-449a-9948-4eee88e83835" />



## Related issue number

Closes #55133

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

kevin85421 commented:
cc @israbbani @codope Would you mind shepherding this PR as core oncalls?

israbbani commented on python/ray/dashboard/modules/job/job_manager.py:
```diff
@@ -163,6 +163,37 @@ async def _monitor_job_internal(
             try:
                 job_status = await self._job_info_client.get_status(job_id)
+
+                if job_supervisor is None:
```
There are two call paths into `_monitor_job_internal`:

1. submit_job (when the job is first submitted). In this case JobSupervisor cannot be None. 
2. _recover_running_jobs which is executed when the JobManager is started up. It looks like this is designed to help with make the JobAgent resilient to restarts. 

Is the bug you've filed in https://github.com/ray-project/ray/issues/55133 related to dashboard restarts? I don't see how you to reproduce the issue without that.

israbbani reviewed (commented):
@daiping8 thank you for the contribution. I think you've found a bug in the recovery path of the JobManager. 

I left a few comments. I'm also reading this code for the first time so lmk if I got something wrong and we can figure it out together.

daiping8 committed:
Add tests for job starting and pending timeouts; handle ping timeouts

- Introduced tests to verify job timeouts during startup and pending states, such as during insufficient resource scenarios or new raylet creation.
- Added timeout handling for job supervisor pings to improve monitoring reliability.

Signed-off-by: daiping8 <dai.ping88@zte.com.cn>

daiping8 committed:
Remove redundant parameters from job timeout tests; comment out unused terminal state logic.

Signed-off-by: daiping8 <dai.ping88@zte.com.cn>

daiping8 committed:
Remove test for pending job timeout during raylet creation; clean up unused logic in `job_manager`.

Signed-off-by: daiping8 <dai.ping88@zte.com.cn>

daiping8 committed:
Add missing `call_ray_start` parameter to `test_job_starting_timeout`.

Signed-off-by: daiping8 <dai.ping88@zte.com.cn>

israbbani commented on python/ray/dashboard/modules/job/job_manager.py:
```diff
@@ -238,7 +246,17 @@ async def _monitor_job_internal(
+                        f"Timeout threshold: {self.PING_TIMEOUT_S} seconds. "
+                        "This may indicate that the cluster does not have enough resources to schedule the job supervisor."
+                    )
+                    continue
```
@daiping8 successfully scheduling the JobSupervisor [requires the resources of the entrypoint for scheduling](https://docs.ray.io/en/latest/cluster/running-applications/job-submission/sdk.html#specifying-cpu-and-gpu-resources). 

When you saw the loop getting blocked at job_supervisor.ping, what resources were you specifying for the entrypoint?

daiping8 replied:
@israbbani    `ray job submit --entrypoint-num-cpus 100 -- python -V`    We submitted a job with an `entrypoint-num-cpus` far exceeding the `num_cpus` in the cluster.

daiping8 committed:
Merge remote-tracking branch 'upstream/master' into dashborad_pending

Signed-off-by: daiping8 <dai.ping88@zte.com.cn>

# Conflicts:
#	python/ray/dashboard/modules/job/job_manager.py
#	python/ray/dashboard/modules/job/tests/test_job_manager.py

daiping8 committed:
Add timeout handling for job supervisor pings and test for job starting timeout.

- Introduced `PING_TIMEOUT_S` to manage ping timeouts for job supervisor.
- Added test to verify job timeout behavior when resources are insufficient.

Signed-off-by: daiping8 <dai.ping88@zte.com.cn>

daiping8 committed:
Merge branch 'master' into dashborad_pending

jjyao commented:
Sorry for the late review. I'll take a look tomorrow.

daiping8 committed:
Merge branch 'master' into dashborad_pending

jjyao commented on python/ray/dashboard/modules/job/job_manager.py:
```diff
@@ -74,6 +74,8 @@ class JobManager:
     LOG_TAIL_SLEEP_S = 1
     JOB_MONITOR_LOOP_PERIOD_S = 1
     WAIT_FOR_ACTOR_DEATH_TIMEOUT_S = 0.1
+    # Timeout duration (in seconds) for job supervisor ping requests.
```
@daiping8 I took a closer look at the code and found the issue is not related to `ping` call.

So we currently have code that is 

```
 try:
                # NOTE: Job monitoring loop sleeps before proceeding with monitoring
                #       sequence to consolidate the control-flow of the pacing
                #       in a single place, rather than having it spread across
                #       many branches
                await asyncio.sleep(self.JOB_MONITOR_LOOP_PERIOD_S)

                job_status = await self._job_info_client.get_status(job_id)
                if job_status == JobStatus.PENDING:
                    # if the actor cannot be started, the job will remaining pending and we check here to see if it times out and fails the task
                    # Compare the current time with the job start time.
                    # If the job is still pending, we will set the status
                    # to FAILED.
                    job_info = await self._job_info_client.get_info(job_id)

                    if time.time() - job_info.start_time / 1000 > timeout:
                                    XXXXX
                    else:
                         continue # THIS IS THE MISSING CODE AND ACTUAL FIX WE NEED, we should `continue` so it goes back to the beginning of the loop and try again, instead of let it run the remaining code which has `ping.remote()` which will hang forever.                
```

daiping8 committed:
Comment out job supervisor ping logic and clean up unused branches in monitoring loop.

Signed-off-by: daiping8 <dai.ping88@zte.com.cn>

jjyao commented on python/ray/dashboard/modules/job/job_manager.py:
```diff
@@ -257,8 +261,18 @@ async def _monitor_job_internal(
 
-                # Verify `JobSupervisor` is alive and reachable
-                await job_supervisor.ping.remote()
+                # # Verify `JobSupervisor` is alive and reachable
+                # try:
+                #     await asyncio.wait_for(
+                #         job_supervisor.ping.remote(), timeout=self.PING_TIMEOUT_S
+                #     )
+                # except asyncio.TimeoutError:
+                #     logger.warning(
+                #         f"Job supervisor ping timed out for job {job_id}. "
+                #         f"Timeout threshold: {self.PING_TIMEOUT_S} seconds. "
+                #         "This may indicate that the cluster does not have enough resources to schedule the job supervisor."
+                #     )
+                #     continue
```
Here we should just keep the old code :)

daiping8 replied:
Yeah. I understand what you mean. I just wanted to verify whether the modifications would pass the CI. In other words, this is still in the draft stage. I didn't expect you to notice it so quickly. Thanks!  ^-^

I've already removed the comments. Now waiting for the CI.

daiping8 committed:
Remove commented-out job supervisor ping logic from monitoring loop.

Signed-off-by: daiping8 <dai.ping88@zte.com.cn>

daiping8 committed:
Reintroduce job supervisor ping logic in monitoring loop.

Signed-off-by: daiping8 <dai.ping88@zte.com.cn>

jjyao committed:
Merge branch 'master' into dashborad_pending

jjyao commented:
@daiping8 could you update the PR title and description to reflect the latest?

daiping8 committed:
Replace `test_job_starting_timeout` with `test_job_timeout_hang_ping_remote` to improve timeout handling test coverage.

Signed-off-by: daiping8 <dai.ping88@zte.com.cn>

daiping8 committed:
Merge remote-tracking branch 'origin/dashborad_pending' into dashborad_pending

daiping8 commented:
> @daiping8 could you update the PR title and description to reflect the latest?

I have done this task, please check it.

daiping8 committed:
Update default `entrypoint_num_cpus` value from 0 to 2 in job manager.

Signed-off-by: daiping8 <dai.ping88@zte.com.cn>

daiping8 commented:
This issue is addressed in a new PR: https://github.com/ray-project/ray/pull/56296. Thanks to @jjyao and @israbbani for their contributions.
